kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,995 @@
 
-use std::sync::Arc;
+use std::{collections::HashSet, sync::Arc};
 
-use crate::{schema::SchemaRef, utils::require, DeltaResult, Error};
+use crate::{
+    schema::{DataType, Schema, SchemaRef, StructField, StructType},
+    DeltaResult, Error,
+};
 
-use arrow_array::RecordBatch;
-use arrow_schema::{Schema as ArrowSchema, SchemaRef as ArrowSchemaRef};
+use arrow_array::{
+    cast::AsArray, new_null_array, Array as ArrowArray, GenericListArray, OffsetSizeTrait,
+    StructArray,
+};
+use arrow_schema::{
+    DataType as ArrowDataType, Field as ArrowField, FieldRef as ArrowFieldRef, Fields,
+    SchemaRef as ArrowSchemaRef,
+};
 use parquet::{arrow::ProjectionMask, schema::types::SchemaDescriptor};
+use tracing::debug;
 
-/// Get the indicies in `parquet_schema` of the specified columns in `requested_schema`. This
-/// returns a tuples of (mask_indicies: Vec<parquet_schema_index>, reorder_indicies:
-/// Vec<requested_index>). `mask_indicies` is used for generating the mask for reading from the
+/// Reordering is specified as a tree. Each level is a vec of `ReorderIndex`s. Each element's index
+/// represents a column that will be in the read parquet data at that level and index. The `index()`
+/// of the element is the position that the column should appear in the final output. If it is a
+/// `Child` variant, then at that index there is a `Struct` whose ordering is specified by the
+/// values in the associated `Vec` according to these same rules.
+#[derive(Debug, PartialEq)]
+pub(crate) enum ReorderIndex {
+    Child {
+        index: usize,
+        children: Vec<ReorderIndex>,
+    },
+    Index {
+        index: usize,
+    },
+    Null {
+        index: usize,
+        field: ArrowFieldRef,
+    },
+}
+
+impl ReorderIndex {
+    fn index(&self) -> usize {
+        match self {
+            ReorderIndex::Child { index, .. } => *index,
+            ReorderIndex::Index { index, .. } => *index,
+            ReorderIndex::Null { index, .. } => *index,
+        }
+    }
+
+    fn set_index(&mut self, target_index: usize) {
+        match self {
+            ReorderIndex::Child { ref mut index, .. } => *index = target_index,
+            ReorderIndex::Index { ref mut index } => *index = target_index,
+            ReorderIndex::Null { ref mut index, .. } => *index = target_index,
+        }
+    }
+
+    /// check if this indexing is ordered. an `Index` variant is ordered by definition. a `Null`
+    /// variant is not because if we have a `Null` variant we need to do work in
+    /// reorder_struct_array to insert the null col
+    fn is_ordered(&self) -> bool {
+        match self {
+            ReorderIndex::Child { ref children, .. } => is_ordered(children),
+            ReorderIndex::Index { .. } => true,
+            ReorderIndex::Null { .. } => false,
+        }
+    }
+}
+
+// count the number of physical columns, including nested ones in an `ArrowField`
+fn count_cols(field: &ArrowField) -> usize {
+    _count_cols(field.data_type())
+}
```

Comment:
Yep, it is. I wrote this inspired by the [`fields`](https://docs.rs/arrow-schema/52.1.0/src/arrow_schema/field.rs.html#362) method in arrow. If it was publicly exposed I'd certainly want a trait, but since this is the only usage I thought the parsimonious (lovely word btw) solution was better.

Reply:
Ack. Especially since the trait wouldn't allow the other simplification below.